src/processors/transform.rs

```diff
@@ -0,0 +1,293 @@
+#[derive(Debug, Copy, Clone)]
+pub struct Transform(pub [[f32; 4]; 4]);
+impl Transform {
+    pub fn identity() -> Self {
+        Transform(
+            [
+                [1.0, 0.0, 0.0, 0.0],
+                [0.0, 1.0, 0.0, 0.0],
+                [0.0, 0.0, 1.0, 0.0],
+                [0.0, 0.0, 0.0, 1.0],
+            ]
+        )
+    }
+}
+
+impl Component for Transform {
+    type Storage = VecStorage<Transform>;
+}
+
+/// Initialization component, added to entity with a `LocalTransform` component after the first update.
+#[derive(Default, Copy, Clone)]
+pub struct Init;
+impl Component for Init {
+    type Storage = NullStorage<Init>;
+}
+
+/// Transformation processor.
+/// Handles updating `Transform` components based on the `LocalTransform` component and parents.
+pub struct TransformProcessor {
+    indices: HashMap<Entity, usize>,
+    sorted: VecDeque<(Entity, Option<Entity>)>,
+}
+
+impl TransformProcessor {
+    pub fn new() -> TransformProcessor {
+        TransformProcessor {
+            indices: HashMap::new(),
+            sorted: VecDeque::new(),
+        }
+    }
+}
+
+impl Processor<Arc<Mutex<Context>>> for TransformProcessor {
+    fn run(&mut self, arg: RunArg, _: Arc<Mutex<Context>>) {
+        let (entities, mut locals, mut globals, mut init) = arg.fetch(|w| {
+            (w.entities(), w.write::<LocalTransform>(), w.write::<Transform>(), w.write::<Init>())
+        });
+
+        let mut new: Vec<Entity> = Vec::new();
+        for (entity, local, _) in (&entities, &locals, !&init).iter() {
+            self.indices.insert(entity, self.sorted.len());
+            self.sorted.push_back((entity, local.parent));
+            new.push(entity.clone());
+        }
+
+        for entity in new {
+            init.insert(entity, Init);
+        }
+
+        // Compute transforms (global) from local transforms and parents.
+        let mut swap_candidates: Vec<(Entity, Entity)> = Vec::new();
+        for &(entity, parent) in self.sorted.iter() {
+            let mut local = locals.get_mut(entity).unwrap();
+
+            if local.dirty {
+                let combined_transform = match parent {
+                    Some(parent) => {
+                        swap_candidates.push((entity, parent));
+
+                        if let Some(parent_global) = globals.get(parent) {
+                            Matrix4::from(parent_global.0) * Matrix4::from(local.matrix())
+                        }
+                        else {
+                            Matrix4::from(local.matrix())
+                        }
+                    },
+                    None => {
+                        Matrix4::from(local.matrix())
+                    },
+                };
+
+                if let Some(global) = globals.get_mut(entity) {
+                    global.0 = combined_transform.into();
+                }
+
+                local.dirty = false;
+            }
+        }
+
+        for (entity, parent) in swap_candidates {

```

Comment:
Right now `swap_candidates` contains a list of entities with parents, then later it iterates over those to check if the child is before the parent. If so then it swaps them in the sorted vec.